File: src/main/java/co/herod/adbwrapper/model/AdbUiNode.java
package co.herod.adbwrapper.model;

import co.herod.adbwrapper.util.UiHierarchyHelper;

public class AdbUiNode {

    private final String nodeString;
    private final long creationTimeMillis;

    private Integer[] bounds;

    public AdbUiNode(final String nodeString) {
        this.nodeString = nodeString;
        this.creationTimeMillis = System.currentTimeMillis();
    }

    public Integer[] getBounds() {
        if (bounds == null || bounds.length != 4) {
            bounds = UiHierarchyHelper.extractBoundsInts(nodeString).blockingSingle();
        }
        return bounds;
    }

    public int getWidth() {
        return UiHierarchyHelper.getWidth(getBounds());
    }

    public int getHeight() {
        return UiHierarchyHelper.getWidth(getBounds());
    }

    public long getCreationTimeMillis() {
        return creationTimeMillis;
    }

    @Override
    public String toString() {
        return "AdbUiNode{" +
                "nodeString='" + nodeString + '\'' +
                '}';
    }

}

File: src/main/java/co/herod/adbwrapper/model/AdbUiNode.kt
package co.herod.adbwrapper.model

import co.herod.adbwrapper.util.UiHierarchyHelper

class AdbUiNode(private val nodeString: String) {

    internal val bounds: Array<Int>? by lazy { UiHierarchyHelper.extractBoundsInts(nodeString)?.blockingSingle() }

    val width: Int? by lazy { bounds?.let { UiHierarchyHelper.getWidth(it) } }

    val height: Int? by lazy { bounds?.let { UiHierarchyHelper.getHeight(it) } }

    val text by lazy { UiHierarchyHelper.extractText(nodeString) }

    override fun toString() = "AdbUiNode{nodeString='$nodeString'}"

}
